### PR TITLE
Pre-compress release files in Brotli before uploading

### PR DESCRIPTION
**Pull request by danielrozenberg**
Partial for https://github.com/ampproject/cdn-worker/pull/91

Tested running this script on my machine and the compression phase pass. I don't have a gcloud upload key locally (nor should I) so the script failed when attempting to actually upload the files, but it looks like the compression part is complete and the upload part hasn't changed

### CODE REVIEW

File: build-system/release-workflows/upload-release.js [hunk 1]
```diff
@@ -100,6 +101,54 @@ function logProgress_(totalFiles, uploadedFiles) {
...
+ * @param {number} sizeHint
+ * @return {Promise<void>}
+ */
+async function brotliCompressFile_(path, sizeHint) {
+  const fileBuffer = fs.readFileSync(path);
+  const compressedBuffer = zlib.brotliCompressSync(fileBuffer, {
+    params: {
+      [zlib.constants.BROTLI_PARAM_QUALITY]: zlib.constants.BROTLI_MAX_QUALITY,
+      [zlib.constants.BROTLI_PARAM_SIZE_HINT]: sizeHint,
+    },
```
Comment (danielrozenberg) [inline build-system/release-workflows/upload-release.js:116]: List of possible params: https://nodejs.org/docs/latest-v16.x/api/zlib.html#compressor-options

File: build-system/release-workflows/upload-release.js [hunk 2]
```diff
@@ -100,6 +101,54 @@ function logProgress_(totalFiles, uploadedFiles) {
...
 
+/**
+ * Compresses a single file with Brotli and writes it to ${path}.br.
+ * @param {string} path
+ * @param {number} sizeHint
+ * @return {Promise<void>}
+ */
+async function brotliCompressFile_(path, sizeHint) {
+  const fileBuffer = fs.readFileSync(path);
+  const compressedBuffer = zlib.brotliCompressSync(fileBuffer, {
```
Comment (jridgewell) [inline build-system/release-workflows/upload-release.js:112]: You're creating an series of promises, but doing all operations sync. We could probably speed up throughput by using streams, eg https://gist.github.com/linoyzaga/e992eecfa371014202cbb8e927935234#file-brotlicompressstream-js.

**Review by jridgewell — APPROVED** [review]
LGTM, deferring to @estherkim

File: build-system/release-workflows/upload-release.js [hunk 3]
```diff
@@ -100,6 +101,54 @@ function logProgress_(totalFiles, uploadedFiles) {
...
+    }
+
+    compressPromises.push(
+      brotliCompressFile_(path, stats.size).then(() => {
+        logProgress_(totalFiles, ++compressedFiles);
+      })
+    );
+  }
+
+  await Promise.all(compressPromises);
```
Comment (estherkim) [inline build-system/release-workflows/upload-release.js:148]: Error handling? Should the CI job fail if compression fails?

File: build-system/release-workflows/upload-release.js (hunk 2 above)
Comment (danielrozenberg) [reply build-system/release-workflows/upload-release.js]: Nice! Significantly faster (<1m compared to >3m) - I love streams!!

File: build-system/release-workflows/upload-release.js (hunk 3 above)
Comment (danielrozenberg) [reply build-system/release-workflows/upload-release.js:156]: I think it should, this is going to become an important part of the CDN soon and it's not a piece of code that's likely to fail often, so if it does it probably means something very bad happened

**Review by estherkim — APPROVED** [review]

**Pull request merged**